llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -365,6 +365,80 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
     EntryPointsGroups[GLOBAL_SCOPE_NAME] = {};
 }
 
+// For device global variables with the 'device_image_scope' property,
+// the function checks that there are no usages of a single device global
+// variable from kernels grouped to different modules.
+void checkImageScopedDeviceGlobals(const Module &M,
+                                   const EntryPointGroupMap &GMap) {
+  // Early exit if there is only one group
+  if (GMap.size() < 2)
+    return;
+
+  // Reverse the EntryPointGroupMap to get a map of entry point -> module's name
+  unsigned EntryPointNumber = 0;
+  for (const auto &Group : GMap) {
+    EntryPointNumber += static_cast<unsigned>(Group.second.size());
+  }
+  DenseMap<const Function *, const StringRef *> EntryPointModules(
+      EntryPointNumber);
+  for (const auto &Group : GMap) {
+    auto *ModuleName = &Group.first;
+    for (const auto *F : Group.second) {
+      EntryPointModules.insert({F, ModuleName});
+    }
+  }
+
+  // Processing device global variables with the "device_image_scope" property
+  for (auto &GV : M.globals()) {
+    if (!isDeviceGlobalVariable(GV) || !hasDeviceImageScopeProperty(GV))
+      continue;
+
+    const StringRef *VarEntryPointModule = nullptr;
+    auto CheckEntryPointModule = [&VarEntryPointModule, &EntryPointModules,
+                                  &GV](const auto *F) {
+      auto EntryPointModulesIt = EntryPointModules.find(F);
+      assert(EntryPointModulesIt != EntryPointModules.end() &&
+             "There is no group for an entry point");
+      if (VarEntryPointModule == nullptr) {
+        VarEntryPointModule = EntryPointModulesIt->second;
+        return;
+      }
+      if (EntryPointModulesIt->second != VarEntryPointModule) {
+        error("device_global variable '" + Twine(GV.getName()) +
+              "' with property \"device_image_scope\" is contained in more "
+              "than one device image.");
+      }
+    };
+
+    SmallVector<const User *, 32> Workqueue;
+    SmallPtrSet<const User *, 32> Visited;
+    for (auto *U : GV.users())
+      Workqueue.push_back(U);
+
+    while (!Workqueue.empty()) {
+      const User *U = Workqueue.back();
+      Workqueue.pop_back();
+      if (!Visited.insert(U).second)
+        continue;
+      if (auto *I = dyn_cast<const Instruction>(U)) {
+        auto *F = I->getFunction();
+        if (!Visited.contains(F))
+          Workqueue.push_back(F);
+      } else {
```

Comment:
I'm not sure it improves readability (this `continue` can easy be overlooked), but because LLVM's codestyle suggests this form of writing, I would not like to argue.